File: data_structure/tree_go/binary_search_tree.go
package tree_go

type BinarySearchTree struct {
	Root *Node
}

func (tree *BinarySearchTree) Insert(key int) {
	if tree.Root == nil {
		tree.Root = &Node{
			Key: key,
		}
		return
	}
	move := tree.Root
	for move != nil {
		if key > move.Key {
			if move.Right == nil {
				move.Right = &Node{Key: key}
				return
			}
			move = move.Right
		} else if key < move.Key {
			if move.Left == nil {
				move.Left = &Node{Key: key}
				return
			}
		} else {
			return
		}
	}
}

func (tree *BinarySearchTree) Find(key int) *Node {
	move := tree.Root
	for move != nil {
		if move.Key == key {
			return move
		}
		if key > move.Key {
			move = move.Right
		} else {
			move = move.Left
		}
	}
	return nil
}

func (tree *BinarySearchTree) Delete(key int) {
	node := tree.Root
	var nodeParent *Node
	for node != nil && node.Key != key {
		nodeParent = node
		if key > node.Key {
			node = node.Right
		} else {
			node = node.Left
		}
	}
	if node == nil {
		return // 没找到
	}

	// 如果删除的节点左右节点都不为空，那么找到右子树中最小节点
	// 将他的数据替换到要删除的节点，然后删除这个柚子属于中最小的节点
	if node.Left != nil && node.Right != nil {
		rightTreeMinNode := node.Right
		rightTreeMinNodeParent := node
		for rightTreeMinNode.Left != nil {
			rightTreeMinNodeParent = rightTreeMinNode
			rightTreeMinNode = rightTreeMinNode.Left
		}
		node.Key = rightTreeMinNode.Key // 替换数据
		node = rightTreeMinNode
		nodeParent = rightTreeMinNodeParent
	}

	// 如果要删除的节点只有左节点，右节点或者是叶子节点
	var child *Node
	if node.Left != nil {
		child = node.Left
	} else if node.Right != nil {
		child = node.Right
	} else {
		child = nil
	}

	// 从树中删除节点
	if nodeParent == nil {
		tree.Root = child
	} else if nodeParent.Left == node {
		nodeParent.Left = child
	} else {
		nodeParent.Right = child
	}
}

func inOrder(node *Node, result []*Node) {
	if node != nil {
		inOrder(node.Left, result)
		result = append(result, node)
		inOrder(node.Right, result)
	}
}

func (tree *BinarySearchTree) InOrder() []*Node {
	keys := make([]*Node, 0)
	inOrder(tree.Root, keys)
	return keys
}

File: data_structure/tree/binary_search_tree.py
# -*- coding:utf-8 -*-

"""
二叉搜索树
@author: 付登龙
@file: binary_search_tree.py
@time: 2019/01/13 10:12 AM
"""
from random import randint, choice
from typing import Optional, List


class TreeNode(object):

    def __init__(self, key: Optional[int]=None, left=None, right=None):
        self.key = key
        self.left = left
        self.right = right


class BinarySearchTree(object):

    def __init__(self, root: Optional[TreeNode] = None):
        self.root = root

    def find(self, key: int) -> Optional[TreeNode]:
        node = self.root
        while node:
            if key > node.key:
                node = node.right
            elif key < node.key:
                node = node.left
            else:
                return node
        return None

    def insert(self, key: int):
        if self.root is None:
            self.root = TreeNode(key)
            return

        node = self.root
        while node is not None:
            if key > node.key:
                if node.right is None:
                    node.right = TreeNode(key)
                    return
                node = node.right
            elif key < node.key:
                if node.left is None:
                    node.left = TreeNode(key)
                    return
                node = node.left
            else:
                return

    def delete(self, key: int):
        node, node_parent = self.root, None

        while node is not None and node.key != key:
            node_parent = node
            if key > node.key:
                node = node.right
            else:
                node = node.left

        if node is None:
            return

        # 要删除的节点有左右子树
        if node.left is not None and node.right is not None:  # 查找右子树最小节点
            right_min_node = node.right
            right_min_node_parent = node
            while right_min_node.left is not None:
                right_min_node_parent = right_min_node
                right_min_node = right_min_node.left
            node.key = right_min_node.key
            node = right_min_node
            node_parent = right_min_node_parent

        # 要删除的节点是叶子节点或者只有一个子节点
        child = None
        if node.left is not None:
            child = node.left
        elif node.right is not None:
            child = node.right

        # 改变要删除节点的父节点指针指向
        if node_parent is None:
            self.root = child
        elif node_parent.left is node:
            node_parent.left = child
        else:
            node_parent.right = child

    def sorted(self) -> List[int]:
        result = []

        def in_order(root: Optional[TreeNode]):
            if root is None:
                return
            in_order(root.left)
            result.append(root.key)
            in_order(root.right)
        in_order(self.root)
        return result


def test():
    binary_search_tree = BinarySearchTree()
    assert len(binary_search_tree.sorted()) == 0

    # test sorted and insert
    for _ in range(40):
        binary_search_tree = BinarySearchTree()
        for _ in range(10):
            binary_search_tree.insert(randint(1, 10 ** 3))
            result = binary_search_tree.sorted()
            assert list(sorted(result)) == result

    # test delete
    for _ in range(40):
        binary_search_tree = BinarySearchTree()
        for _ in range(10):
            binary_search_tree.insert(randint(1, 10 ** 3))
        result = binary_search_tree.sorted()
        binary_search_tree.delete(choice(result))
        assert list(sorted(result)) == result

    # test find
    for _ in range(40):
        binary_search_tree = BinarySearchTree()
        for _ in range(10):
            binary_search_tree.insert(randint(1, 10 ** 3))
        result = binary_search_tree.sorted()
        assert binary_search_tree.find(choice(result)) is not None
        assert binary_search_tree.find(randint(10 ** 4, 10 ** 6)) is None

    # test operation on empty tree
    binary_search_tree = BinarySearchTree()
    assert binary_search_tree.find(0) is None
    binary_search_tree.insert(0)
    assert binary_search_tree.find(0) is not None
    binary_search_tree.delete(0)
    assert binary_search_tree.find(0) is None
